common/coins.go

```diff
@@ -80,3 +77,76 @@ func (a EtherAmount) String() string {
 func FmtFloat(f float64) string {
 	return strconv.FormatFloat(f, 'f', -1, 64)
 }
+
+// ERC20TokenAmount represents some amount of an ERC20 token in the smallest denomination
+type ERC20TokenAmount struct {
+	amount   *big.Int
+	numUnits float64 // 10^decimals
+}
+
+// NewERC20TokenAmountFromBigInt converts some amount in the smallest token denomination into an ERC20TokenAmount.
+func NewERC20TokenAmountFromBigInt(amount *big.Int, decimals int) *ERC20TokenAmount {
+	return &ERC20TokenAmount{
+		amount:   amount,
+		numUnits: math.Pow10(decimals),
+	}
+}
+
+// NewERC20TokenAmount converts some amount in the smallest token denomination into an ERC20TokenAmount.
+func NewERC20TokenAmount(amount int64, decimals int) *ERC20TokenAmount {
+	return &ERC20TokenAmount{
+		amount:   big.NewInt(amount),
+		numUnits: math.Pow10(decimals),
+	}
+}
+
+// NewERC20TokenAmountFromDecimals converts some amount of standard token in standard format
+// to its smaller denomination.
+// For example, if amount is 1.99 and decimals is 9, the resulting value stored
+// is 1.99 * 10^9.
+func NewERC20TokenAmountFromDecimals(amount float64, decimals int) *ERC20TokenAmount {
+	numUnits := math.Pow10(decimals)
+	amt := big.NewFloat(amount)
+	mult := big.NewFloat(numUnits)
+	prod := new(big.Float).Mul(amt, mult)
+	res := round(prod)
+	return &ERC20TokenAmount{
+		amount:   res,
+		numUnits: numUnits,
+	}
+}
+
+// BigInt returns the given ERC20TokenAmount as a *big.Int
+func (a *ERC20TokenAmount) BigInt() *big.Int {
+	return a.amount
+}
+
+// AsStandard returns the amount in standard form
+func (a *ERC20TokenAmount) AsStandard() float64 {
+	wei := new(big.Float).SetInt(a.BigInt())
+	mult := big.NewFloat(a.numUnits)
+	ether := new(big.Float).Quo(wei, mult)
+	res, _ := ether.Float64()
+	return res
+}
+
+// String ...
+func (a *ERC20TokenAmount) String() string {
+	return a.BigInt().String()
+}
+
+// round rounds the input *big.Float to a *big.Int.
+// eg. if the input is 33.49, it returns 33.
+// if the input is 33.5, it returns 34.
+func round(num *big.Float) *big.Int {
```

Comment:
There's another possible algorithm here if you like it better (not saying it is better and I didn't test it): https://stackoverflow.com/a/60810320/2049639